src/java.base/share/classes/sun/security/provider/NamedSignature.java

```diff
@@ -0,0 +1,222 @@
+        } else {
+            throw new IllegalStateException("No private key");
+        }
+    }
+
+    @Override
+    protected boolean engineVerify(byte[] sig) throws SignatureException {
+        if (pubKey != null) {
+            var msg = bout.toByteArray();
+            bout.reset();
+            return implVerify(name, pubKey, pk2, msg, sig);
+        } else {
+            throw new IllegalStateException("No public key");
+        }
+    }
+
+    @Override
+    @SuppressWarnings("deprecation")
+    protected void engineSetParameter(String param, Object value)
+            throws InvalidParameterException {
+        throw new UnsupportedOperationException("setParameter() not supported");
+    }
+
+    @Override
+    @SuppressWarnings("deprecation")
+    protected Object engineGetParameter(String param) throws InvalidParameterException {
+        throw new UnsupportedOperationException("getParameter() not supported");
+    }
+
+    @Override
+    protected void engineSetParameter(AlgorithmParameterSpec params)
+            throws InvalidAlgorithmParameterException {
+        if (params != null) {
+            throw new InvalidAlgorithmParameterException(
+                    "The " + fname + " algorithm does not take any parameters");
+        }
+    }
+
+    @Override
+    protected AlgorithmParameters engineGetParameters() {
+        return null;
+    }
+
+    /// User-defined sign function.
+    ///
+    /// @param name parameter name
+    /// @param sk private key in raw bytes
+    /// @param sk2 parsed private key, `null` if none.
+    /// @param msg the message
+    /// @param sr SecureRandom object, `null` if not initialized
+    /// @return the signature
+    /// @throws ProviderException if there is an internal error
+    /// @throws SignatureException if there is another error
+    protected abstract byte[] implSign(String name, byte[] sk, Object sk2,
+            byte[] msg, SecureRandom sr) throws SignatureException;
+
+    /// User-defined verify function.
+    ///
+    /// @param name parameter name
+    /// @param pk public key in raw bytes
+    /// @param pk2 parsed public key, `null` if none.
+    /// @param msg the message
+    /// @param sig the signature
+    /// @return true if verified
+    /// @throws ProviderException if there is an internal error
+    /// @throws SignatureException if there is another error
+    protected abstract boolean implVerify(String name, byte[] pk, Object pk2,
+            byte[] msg, byte[] sig) throws SignatureException;
+
+    /// User-defined function to validate a public key.
+    ///
+    /// This method will be called in `initVerify`. This gives provider a chance to
+    /// reject the key so an `InvalidKeyException` can be thrown earlier.
+    /// An implementation can optional return a "parsed key" as an `Object` value.
+    /// This object will be passed into the [#implVerify] method along with the raw key.
+    ///
+    /// The default implementation returns `null`.
```

Comment:
If the majority of implementations are supposed to implement this method, then it may be better to make it abstract (so that it won't be accidentally missed) and add a note that subclasses should return null if they don't have any additional checks.

Reply:
I'll discuss my customers.

Reply:
After some discussion, we decided to keep the current design. I acknowledge that there are advantages to override these methods, and it’s even encouraged. However, considering the nature of these methods, they can be omitted without impacting functionality. Finally, these are internal classes and we have controls.